File: hash-table/hash_table.clj
(defn make-hash-table [size]
  {:size size :buckets (vec (repeat size []))})

(defn hash-fn [key size]
  (mod (hash key) size))

(defn put [ht key value]
  (let [index (hash-fn key (:size ht))
        bucket (nth (:buckets ht) index)
        new-bucket (conj (remove #(= (first %) key) bucket) [key value])]
    (assoc-in ht [:buckets index] new-bucket)))

(defn get-val [ht key]
  (let [index (hash-fn key (:size ht))
        bucket (nth (:buckets ht) index)
        pair (first (filter #(= (first %) key) bucket))]
    (when pair (second pair))))

(def ht (-> (make-hash-table 10)
            (put "key1" "value1")))
(println (get-val ht "key1"))

File: hash-table/hash_table.js
class HashTable {
    constructor(size = 10) {
        this.size = size;
        this.table = Array(size).fill(null).map(() => []);
    }
    
    _hash(key) {
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash = (hash + key.charCodeAt(i) * i) % this.size;
        }
        return hash;
    }
    
    put(key, value) {
        const index = this._hash(key);
        const bucket = this.table[index];
        
        for (let i = 0; i < bucket.length; i++) {
            if (bucket[i][0] === key) {
                bucket[i][1] = value;
                return;
            }
        }
        bucket.push([key, value]);
    }
    
    get(key) {
        const index = this._hash(key);
        const bucket = this.table[index];
        
        for (const [k, v] of bucket) {
            if (k === key) return v;
        }
        throw new Error(`Key ${key} not found`);
    }
}

const ht = new HashTable();
ht.put("name", "Alice");
ht.put("age", 30);
console.log(ht.get("name")); // Alice
